### PR TITLE
Update to ChainRulesCore 0.5.1

### PR DESCRIPTION
**Pull request by YingboMa**

### CODE REVIEW

**Comment by nickrobinson251** [comment]
Maybe point this PR at Lyndon's branch? https://github.com/JuliaDiff/ChainRules.jl/pull/145

**Comment by YingboMa** [comment]
Waiting for https://github.com/JuliaDiff/ChainRulesCore.jl/pull/90

File: test/rulesets/Base/base.jl [hunk 1]
```diff
@@ -72,31 +91,11 @@
...
-                # this check is needed because these have discontinuities between
-                # `-10 + im*eps()` and `-10 - im*eps()`
-                should_test_wirtinger = imag(x) != 0 && real(x) < 0
-                test_scalar(sqrt, x; test_wirtinger=should_test_wirtinger)
-                test_scalar(log, x; test_wirtinger=should_test_wirtinger)
-                test_scalar(log2, x; test_wirtinger=should_test_wirtinger)
-                test_scalar(log10, x; test_wirtinger=should_test_wirtinger)
-                test_scalar(log1p, x; test_wirtinger=should_test_wirtinger)
-            end
-        end
```
Comment (oxinabox) [inline test/rulesets/Base/base.jl]: These shouldn't have been deleted.
Just replaced with:

```
test_scalar(cbrt, x)
if x >= 0
    test_scalar(sqrt, x)
    test_scalar(log, x)
    test_scalar(log2, x)
    test_scalar(log10, x)
    test_scalar(log1p, x)
end
```

File: test/rulesets/Base/base.jl [hunk 2]
```diff
@@ -72,31 +91,11 @@
...
-    @testset "Unary complex functions" begin
-        for x in (-4.1, 6.4, 1.0+0.5im, -10.0+1.5im)
-            test_scalar(real, x)
-            test_scalar(imag, x)
-
-            test_scalar(abs, x)
-            test_scalar(hypot, x)
-
-            test_scalar(angle, x)
-            test_scalar(abs2, x)
```
Comment (oxinabox) [inline test/rulesets/Base/base.jl]: These shouldn't have been deleted,
only the cases where they are not holomorphic should have been deleted.

We definately still need tests of `abs` and `hypot` for real values.

File: test/rulesets/Base/base.jl [hunk 3]
```diff
@@ -72,31 +91,11 @@
...
             test_scalar(conj, x)
+            test_scalar(adjoint, x)
+            test_scalar(abs2, x)
+
+            x isa Real && test_scalar(cbrt, x)
```
Comment (oxinabox) [inline test/rulesets/Base/base.jl]:
```suggestion
            test_scalar(cbrt, x)
```

File: test/test_util.jl [hunk 4]
```diff
@@ -16,50 +16,25 @@ at input point `x` to confirm that there are correct ChainRules provided.
...
-    @testset "$f at $x, $(nameof(rule))" for rule in (rrule, frule)
-        res = rule(f, x)
-        @test res !== nothing  # Check the rule was defined
-        fx,  prop_rule = res
+    r_res = rrule(f, x)
+    f_res = frule(f, x, Zero(), 1)
+    @test r_res !== f_res !== nothing  # Check the rule was defined
+    r_fx, prop_rule = r_res
+    f_fx, f_∂x = f_res
+    @testset "$f at $x, $(nameof(rule))" for (rule, fx, ∂x) in ((rrule, r_fx, prop_rule(1)), (frule, f_fx, f_∂x))
```
Comment (oxinabox) [inline test/test_util.jl]: Line is overlength.

```suggestion
    @testset "$f at $x, $(nameof(rule))" for (rule, fx, ∂x) in (
        (rrule, r_fx, prop_rule(1)),
        (frule, f_fx, f_∂x)
    )
```

File: test/test_util.jl [hunk 5]
```diff
@@ -16,50 +16,25 @@ at input point `x` to confirm that there are correct ChainRules provided.
...
-            @test isapprox(
-                wirtinger_primal(∂x), ∂;
-                rtol=rtol, atol=atol, kwargs...
-            )
-            @test isapprox(
-                wirtinger_conjugate(∂x), ∂̅;
-                rtol=rtol, atol=atol, kwargs...
-            )
         end
+        @test isapprox(∂x, fdm(f, x);
```
Comment (oxinabox) [inline test/test_util.jl]: If you are going to roll this up onto one linem, do it fully.
It fits its under the limit:

```suggestion
        @test isapprox(∂x, fdm(f, x); rtol=rtol, atol=atol, kwargs...)
```

File: test/test_util.jl [hunk 6]
```diff
@@ -16,50 +16,25 @@ at input point `x` to confirm that there are correct ChainRules provided.
...
+                       rtol=rtol, atol=atol, kwargs...)
```
Comment (oxinabox) [inline test/test_util.jl]:
```suggestion
```

File: src/rulesets/LinearAlgebra/blas.jl [hunk 7]
```diff
@@ -70,11 +67,8 @@ end
 ##### `BLAS.asum`
 #####
 
-function frule(::typeof(BLAS.asum), x)
-    function asum_pushforward(_, Δx)
-        return sum(cast(sign, x) * Δx)
-    end
-    return BLAS.asum(x), asum_pushforward
+function frule(::typeof(BLAS.asum), x, _, Δx)
+    return BLAS.asum(x), sum(cast(sign, x) * Δx)
```
[the remark this answers is not in the document]
Comment (YingboMa) [reply src/rulesets/LinearAlgebra/blas.jl]: There is actually no `BLAS.asum(x)`, so I removed it.

[the remark this answers is not in the document]
Comment (oxinabox) [reply src/rulesets/LinearAlgebra/blas.jl]: Fair enough, wonder how it got in there,

[the remark this answers is not in the document]
Comment (YingboMa) [reply src/rulesets/LinearAlgebra/blas.jl]: Ah, I confused myself. `BLAS.asum(x)` is there, but it is not documented in Julia
```julia
help?> BLAS.asum
  asum(n, X, incx)

  Sum of the absolute values of the first n elements of array X with stride incx.

  Examples
  ≡≡≡≡≡≡≡≡≡≡

  julia> BLAS.asum(5, fill(1.0im, 10), 2)
  5.0

  julia> BLAS.asum(2, fill(1.0im, 10), 5)
  2.0
```

[the remark this answers is not in the document]
Comment (YingboMa) [reply src/rulesets/LinearAlgebra/blas.jl]: Added back.

**Pull request merged**